# Day 1:- 11/7/2026

# 1. Write a program to print "Hello, World!" on the screen.

In [1]:
print("Hello, World!")

Hello, World!


# 2. Write a program to read two numbers and print their sum.

In [3]:
num1 = int(input("Enter the first number : "))
num2 = int(input("Enter the second number : "))
print(f"Sum of {num1} and {num2} = {num1 + num2}")

Enter the first number :  5
Enter the second number :  6


Sum of 5 and 6 = 11


# 3. Write a program to read two numbers and print their sum, difference, product and quotient.

In [7]:
num1 = int(input("Enter the first number : "))
num2 = int(input("Enter the second number : "))
print(f"""Sum of {num1} and {num2} = {num1 + num2}
Difference of {num1} and {num2} = {num1 - num2}
Product of {num1} and {num2} = {num1 * num2}
Quotient of {num1} and {num2} = {num1 / num2}""")

Enter the first number :  100
Enter the second number :  2


Sum of 100 and 2 = 102
Difference of 100 and 2 = 98
Product of 100 and 2 = 200
Quotient of 100 and 2 = 50.0


# 4. Write a program to read the radius of a circle and print its area and circumference.

In [21]:
from math import pi

PI = float(f"{pi:.4f}") # This takes only the first 4 values after the decimal point in the value of pi
radius = float(input("Enter the radius of a circle: "))
print(f"""Area of Circle = {PI * ( radius ** 2)} sq. units
Circumference of Circle = {2 * PI * radius} units""")


Enter the radius of a circle:  100


Area of Circle = 31416.0 sq. units
Circumference of Circle = 628.3199999999999 units


# 5. Write a program to read the length and breadth of a rectangle and print its area and perimeter.

In [22]:
length = float(input("Enter the length : "))
breadth = float(input("Enter the breadth: "))
print(f"""Area of rectangle = {length * breadth} sq. units
Perimeter of rectangle = {2 * (length + breadth)} units""")

Enter the length :  10
Enter the breadth:  20


Area of rectangle = 200.0 sq. units
Perimeter of rectangle = 60.0 units


# 6. Write a program to swap two numbers using a third variable.

In [23]:
num1 = int(input("Enter the first number : "))
num2 = int(input("Enter the second number: "))
print(f"Numbers before Swapping: {num1} & {num2}")
temp = num2
num2 = num1
num1 = temp
print(f"Numbers after Swapping: {num1} & {num2}")

Enter the first number :  6
Enter the second number:  9


Numbers before Swapping: 6 & 9
Numbers after Swapping: 9 & 6


# 7. Write a program to swap two numbers without using a third variable.

In [29]:
# n1 = 6
# n2 = 9
# n1 = n1 + n2 = 6 + 9 = 15
# n2 = n1 - n2  =  15 - 9 = 6
# n1 = n1 - n2 = 15 - 6 = 9

num1 = int(input("Enter the first number : "))
num2 = int(input("Enter the second number : "))
print(f"Numbers before swapping: num1={num1} & num2={num2}")
num1 = num1 + num2
num2 = num1 - num2
num1 = num1 - num2
print(f"Numbers after swapping: num1={num1} & num2={num2}")

Enter the first number :  6
Enter the second number :  9


Numbers before swapping: num1=6 & num2=9
Numbers after swapping: num1=9 & num2=6


# 8. Write a program to read a temperature in Celsius and convert it to Fahrenheit.

In [32]:
# deg F = deg C * (9/5)  + 32
deg_C = float(input("Enter the temperature in degree Celsius: "))
deg_F = (deg_C * 9) / 5 + 32
print(f"{deg_C} degree Celsius = {deg_F} degree Fahrenheit.")

Enter the temperature in degree Celsius:  5


5.0 degree Celsius = 41.0 degree Fahrenheit.


# 9. Write a program to read the marks of 5 subjects and print the total and average.

In [24]:
marks1 = float(input("Enter the marks of Subject 1: "))
marks2 = float(input("Enter the marks of Subject 2: "))
marks3 = float(input("Enter the marks of Subject 3: "))
marks4 = float(input("Enter the marks of Subject 4: "))
marks5 = float(input("Enter the marks of Subject 5: "))
total_marks = marks1 + marks2 + marks3 + marks4 + marks5
avg_marks = total_marks / 5
print(f"Total marks = {total_marks}\nAverage_marks = {avg_marks}")

Enter the marks of Subject 1:  10
Enter the marks of Subject 2:  20
Enter the marks of Subject 3:  30
Enter the marks of Subject 4:  40
Enter the marks of Subject 5:  50


Total marks = 150.0
Average_marks = 30.0


# 10. Write a program to read seconds and convert them into hours, minutes and seconds.

In [27]:
seconds = int(input("Enter seconds to convert them to Hour: Minutes: Seconds format: "))
hours = seconds // (60 * 60)
minutes = (seconds - hours * 60 * 60 ) // 60
secs = seconds - hours * 60 * 60 - minutes * 60 
print(f"{seconds} in Hours: Minutes: seconds = {hours}: {minutes}: {secs}")


Enter seconds to convert them to Hour: Minutes: Seconds format:  12352


12352 in Hours: Minutes: seconds = 3: 25: 52


# Day 2: 12/07/2026

# 11. Write a program to read a number and check whether it is even or odd.





In [3]:
num = int(input("Enter a number to check if it is Even or Odd: "))
if num % 2 != 0:
    print(f"{num} is an Odd number.")
else:
    print(f"{num} is an Even number.")

Enter a number to check if it is Even or Odd:  53


53 is an Odd number.


In [5]:
num = int(input("Enter the number to check if it is Even or Odd: "))
print(f"{num} is an Odd number.") if num % 2 != 0 else print(f"{num} is an Even number.")

Enter the number to check if it is Even or Odd:  59


59 is an Odd number.


# 12. Write a program to read a number and check whether it is positive, negative or zero.

In [8]:
num = int(input("Enter the number to check if it is +/-/0: "))
if num > 0:
    print(f"{num} is positive number.")
elif num < 0: 
    print(f"{num} is negative number.")
else:
    print(f"{num} is a zero.")

Enter the number to check if it is +/-/0:  0


0 is a zero.


# 13. Write a program to read three numbers and find the largest among them.

In [11]:
num1 = float(input("Enter the first number: "))
num2 = float(input("Enter the second number: "))
num3 = float(input("Enter the third number: "))

if num1 >= num2:
    if num2 >= num3:
        print(f'num1 {num1} is the largest number.')
    elif num1 >= num3:
        print(f'num1 {num1} is the largest number.')
    else: 
        print(f"num3 {num3} is the largest number.")
elif num2 >= num3:
    print(f"num2 {num2} is the largest number.")
else:
    print(f"num3 {num3} is the largest number.")

Enter the first number:  98
Enter the second number:  99
Enter the third number:  100


num3 100.0 is the largest number.


In [14]:
num1 = int(input("Enter the first number: "))
num2 = int(input("Enter the second number: "))
num3 = int(input("Enter the third number: "))

if num1 >= num2 and num1 >= num3:
    print(f"num1 {num1} is the largest number.")
elif num2 >= num3:
    print(f"num2 {num2} is the largest number.")
else: 
    print(f"num3 {num3} is the largest number.")

Enter the first number:  56
Enter the second number:  52
Enter the third number:  86


num3 86 is the largest number.


# 14. Write a program to read three numbers and find the smallest among them.

In [20]:
num1 = int(input("Enter the first number: "))
num2 = int(input("Enter the second number: "))
num3 = int(input("Enter the third number: "))

if num1 <= num2 and num1 <= num3:
    print(f"num1 {num1} is the smallest number.")
elif num2 <= num3:
    print(f"num2 {num2} is the smallest number.")
else:
    print(f"num3 {num3} is the smallest number.")

Enter the first number:  -15
Enter the second number:  -12
Enter the third number:  -1


num1 -15 is the smallest number.


# 15. Write a program to read a year and check whether it is a leap year or not.

In [14]:
# logic => year % 4 == 0 and  year % 100 != 0 OR year % 400 == 0 is a leap year else not a leap year
year = int(input("Enter the year to check whether it is a leap year or not: "))
if (year % 4 == 0 and year % 100 != 0) or (year % 400 == 0):
    print(f"{year} is a leap year.")
else:
    print(f"{year} is NOT a leap year.")

Enter the year to check whether it is a leap year or not:  1900


1900 is NOT a leap year.


# 16. Write a program to read a character and check whether it is a vowel or a consonant.

In [4]:
char = input("Enter the character to check whether it is a vowel or a consonant: ")
vowels = ['a','e','i','o','u','A','E','I','O','U']
if char in vowels:
    print(f"Character '{char}' is a vowel.")
else:
    print(f"Character '{char}' is a consonant.")

Enter the character to check whether it is a vowel or a consonant:  U


Character 'U' is a vowel.


# 17. Write a program to read a character and check whether it is an alphabet, digit or special symbol.

In [12]:
char = input('Enter the character to check whether it is an alphabet/digit/special symbol: ')
if len(char) == 1:
    if char.isalpha():
        print(f"Character '{char}' is an alphabet.")
    elif char.isdigit():
        print(f"Character '{char}' is a digit.")
    else:
        print(f"Character '{char}' is a special symbol.")
else:
    print("Please enter exactly one character.")

Enter the character to check whether it is an alphabet/digit/special symbol:  g


Character 'g' is an alphabet.


# 18. Write a program to read the marks of a student and print the grade (A/B/C/D/Fail).

Assumption:

    marks range    Grade
    91 - 100          A
    71 - 90           B
    51 - 70           C
    31 - 50           D
    0 - 30           Fail

In [1]:
marks = float(input("Enter the marks of student to corresponding grades: "))
if 90 < marks <= 100:
    print("Grade is 'A'")
elif 70 < marks <=90:
    print("Grade is 'B'")
elif 50 < marks <= 70:
    print("Grade is 'C'")
elif 30 < marks <= 50:
    print("Grade is 'D'")
else:
    print("Fail")

Enter the marks of student to corresponding grades:  86


Grade is 'B'


# 19. Write a program to read a number and check whether it is divisible by both 3 and 5.

In [29]:
num = int(input("Enter the number to check its divisibility with both 3 and 5: "))
if num % 3 == 0 and num % 5 == 0:
    print(f"Number {num} is divisible by both 3 and 5.")
else: 
    print(f"Number {num} is NOT divisible by both 3 and 5.")

Enter the number to check its divisibility with both 3 and 5:  45


Number 45 is divisible by both 3 and 5.


In [31]:
num = int(input("Enter the number to check its divisibility by both 3 and 5: "))

if num % 15 == 0:
    print(f"Number {num} is divisible by both 3 and 5.")
else:
    print(f"Number {num} is NOT divisible by both 3 and 5.")

Enter the number to check its divisibility by both 3 and 5:  26


Number 26 is NOT divisible by both 3 and 5.


# 20. Write a program to read the age of a person and check whether they are eligible to vote.

In [26]:
age = int(input("Enter your age to check if you're eligible to vote or not: "))
if age > 0:
    if age >= 18:
        print(f"With the age {age} you are eligible to vote.")
    else:
        print(f"With the age {age} you are NOT eligible to vote.")
else:
    raise ValueError("Enter valid age. Age can not be <= 0.")

Enter your age to check if you're eligible to vote or not:  18


With the age 18 you are eligible to vote.


# Day 3: 15/07/2026

# 21. Write a program to display all the natural numbers from 1 to n. (n is user input)

In [1]:
n = int(input("Enter a number 'n': "))
for num in range(1, n + 1):
    print(num, end=", ")

Enter a number 'n':  5


1, 2, 3, 4, 5, 

# 22. Write a program to display all natural numbers from 1 to n in reverse order.

In [3]:
n = int(input("Enter the number 'n': "))
for num in range(n, 0, -1):
    print(num, end=", ")

Enter the number 'n':  5


5, 4, 3, 2, 1, 

# 23. Write a program to display all even numbers from 1 to n.

In [6]:
n = int(input("Enter the number 'n': "))
print(f"Displaying all even numbers between 1 and {n}...")
for num in range(1, n + 1):
    if num % 2 == 0 :
        print(num, end=", ") 

Enter the number 'n':  20


Displaying all even numbers between 1 and 20...
2, 4, 6, 8, 10, 12, 14, 16, 18, 20, 

# 24. Write a program to display all odd numbers from 1 to n.

In [7]:
n = int(input("Enter the number 'n': "))
print(f"Displaying all Odd numbers between 1 and {n}...")
for num in range(1, n + 1):
    if num % 2 != 0:
        print(num, end=", ")

Enter the number 'n':  20


Displaying all Odd numbers between 1 and 20...
1, 3, 5, 7, 9, 11, 13, 15, 17, 19, 

# 25. Write a program to find the sum of all natural numbers from 1 to n.

In [8]:
n = int(input("Enter the number 'n': "))
total = 0
for num in range(1, n+1):
    total += num
print(f"Sum of all natural numbers from 1 to {n} = {total}")

Enter the number 'n':  10


Sum of all natural numbers from 1 to 10 = 55


In [10]:
n = int(input("Enter the number 'n': "))
print(f"Sum of all natural numbers from 1 to {n} = {int((n*(n+1))/2)}")

Enter the number 'n':  10


Sum of all natural numbers from 1 to 10 = 55


# Day 4: 16/07/2026

# 26. Write a program to find the sum of all even numbers from 1 to n.

In [3]:
n = int(input("Enter the number 'n': "))
even_sum = 0
for num in range(1, n + 1):
    if num % 2 == 0:
        even_sum += num

print(f"Sum of all even numbers from 1 to {n} is {even_sum}.")

Enter the number 'n':  10


Sum of all even numbers from 1 to 10 is 30.


In [2]:
n = int(input("Enter the number 'n': "))
no_of_even_num = n // 2
print(f"Sum of all even numbers from 1 to {n} is {no_of_even_num * (no_of_even_num + 1)}")

Enter the number 'n':  10


Sum of all even numbers from 1 to 10 is 30


# 27. Write a program to find the sum of all odd numbers from 1 to n.

In [6]:
# To get sum of all odd numbers = Sum of first all numbers - Sum of all even numbers
n = int(input("Enter the number 'n': "))
no_of_evens = n // 2
print(f"Sum of all odd numbers from 1 to {n} = {(n * (n+1) * 0.5) - no_of_even_num * (no_of_even_num + 1)}")

Enter the number 'n':  11


Sum of all odd numbers from 1 to 11 = 36.0


In [7]:
n = int(input("Enter the number 'n': "))
odd_sum = 0
for num in range(1, n+1, 2):
    if num % 2 != 0:
        odd_sum += num

print(f"Sum of all odd numbers from 1 to {n} = {odd_sum}")

Enter the number 'n':  11


Sum of all odd numbers from 1 to 11 = 36


# 28. Write a program to find the product of all natural numbers from 1 to n (factorial of n).

In [8]:
n = int(input("Enter the number 'n': "))
factorial = 1
for num in range(1, n + 1):
    factorial *= num

print(f"Product of all natural numbers from 1 to {n} = {factorial}")

Enter the number 'n':  5


Product of all natural numbers from 1 to 5 = 120


# Day 5: 19/07/2026

# 29. Write a program to display the multiplication table of a number n.

In [3]:
num = int(input("Enter the number 'n' of which you want multiplication table: "))
for i in range(1, 11):
    print(f"{num} X {i} = {num * i}") 

Enter the number 'n' of which you want multiplication table:  5


5 X 1 = 5
5 X 2 = 10
5 X 3 = 15
5 X 4 = 20
5 X 5 = 25
5 X 6 = 30
5 X 7 = 35
5 X 8 = 40
5 X 9 = 45
5 X 10 = 50


# 30. Write a program to display all multiples of a number m up to n terms.

In [16]:
m, n = (int(input("Enter a number 'm' : ")), int(input("Enter the number of terms 'n' : ")))
print(f"{n} multiples of {m} are:")
for i in range(1, n+1):
    print(m * i)

Enter a number 'm' :  10
Enter the number of terms 'n' :  12


12 multiples of 10 are:
10
20
30
40
50
60
70
80
90
100
110
120


# 31. Write a program to count how many numbers from 1 to n are divisible by 3.

In [20]:
num = int(input("Enter a number: "))
count = 0
result = []
for i in range(1, num + 1):
    if i % 3 == 0:
        result.append(i)
        count += 1

print(f"There are {count} numbers {result} from 1 to {num} which are divisble by 3.")

Enter a number:  15


There are 5 numbers [3, 6, 9, 12, 15] from 1 to 15 which are divisble by 3.


# 32. Write a program to display all numbers from 1 to n that are divisible by 3 or 5.

In [23]:
n = int(input("Enter a number: "))
numbers = []
for i in range(1, n + 1):
    if i % 3 == 0 or i % 5 == 0:
        numbers.append(i)

print(f"Numbers from 1 to {n} that are divisible by 3 or 5 are : ")
for num in numbers:
    print(num, end=' ')

Enter a number:  100


Numbers from 1 to 100 that are divisible by 3 or 5 are : 
3 5 6 9 10 12 15 18 20 21 24 25 27 30 33 35 36 39 40 42 45 48 50 51 54 55 57 60 63 65 66 69 70 72 75 78 80 81 84 85 87 90 93 95 96 99 100 

In [24]:
n = int(input("Enter a number: "))
for i in range(1, n + 1):
    if i % 3 == 0 or i % 5 == 0:
        print(i, end=" ")

Enter a number:  100


3 5 6 9 10 12 15 18 20 21 24 25 27 30 33 35 36 39 40 42 45 48 50 51 54 55 57 60 63 65 66 69 70 72 75 78 80 81 84 85 87 90 93 95 96 99 100 

# 33. Write a program to count the number of digits in a number n.

In [25]:
number = int(input("Enter the number: "))
count = 0
temp = number
while temp > 0:
    temp = temp // 10
    count += 1
print(f"There are {count} digits in {number}")

Enter the number:  12345


There are 5 digits in 12345


In [27]:
number = input("Enter the number: ")
count = 0
for digit in number:
    count += 1

print(f"There are {count} digits in {int(number)}")

Enter the number:  1234568


There are 7 digits in 1234568


In [28]:
number = int(input("Enter the number: "))
number = abs(number) # handles negative numbers
print(f"Numbers of digits : {len(str(number))}")

Enter the number:  12564789


Numbers of digits : 8


In [29]:
# production ready code
def count_digits(n: int) -> int:
    # handles negative numbers
    n = abs(n)

    # special case for 0
    if n == 0:
        return 1

    count = 0
    while n > 0:
        n //= 10
        count += 1
    return count

number = -123523
print(f"Number of digits : {count_digits(number)}")

Number of digits : 6


# 34. Write a program to display all the digits of a number n (one per line).

In [30]:
number = int(input("Enter the number: "))
while number > 0:
    digit = number % 10
    number //= 10
    print(digit)

Enter the number:  123


3
2
1


In [34]:
number = int(input("Enter the number: "))
digits = []
while number > 0:
    digit = number % 10
    digits.append(digit)
    number //= 10
for digit in reversed(digits):
    print(digit)

Enter the number:  123456


1
2
3
4
5
6


In [40]:
number = int(input("Enter the number: "))
digits = []
while number > 0:
    digit = number % 10
    digits.append(digit)
    number //= 10
for idx in range(len(digits) - 1,-1,-1):
    print(digits[idx])

Enter the number:  123456


1
2
3
4
5
6


In [41]:
number = input("Enter the number: ")
for digit in number:
    print(digit)

Enter the number:  123456


1
2
3
4
5
6


In [56]:
number = int(input("Enter the number: "))
num_of_digits = count_digits(number)
while num_of_digits > 0:
    digit = number // (10 ** (num_of_digits - 1))
    print(digit)
    number %= (10 ** (num_of_digits -1))
    num_of_digits -= 1

Enter the number:  123456


1
2
3
4
5
6


In [57]:
number = int(input("Enter the number: "))
num_of_digits = len(str(number))
while num_of_digits > 0:
    digit = number // (10 ** (num_of_digits - 1))
    print(digit)
    number %= (10 ** (num_of_digits -1))
    num_of_digits -= 1

Enter the number:  12345


1
2
3
4
5


# 35. Write a program to find the sum of all digits of a number n.

In [42]:
def digits_sum(n: int) -> int:
    n = abs(n)

    total = 0
    while n > 0:
        digit = n % 10 # last digit is fetched first
        total += digit
        n //= 10
    return total

num = -1234
print(f"Sum of all digits of {num}: {digits_sum(num)}")

Sum of all digits of -1234: 10


In [59]:
def digits_sum2(n: int) -> int:
    n = abs(n)

    total = 0
    num_of_digits = len(str(n))
    while num_of_digits > 0:
        digit = n // (10 ** (num_of_digits - 1))
        total += digit
        n %= (10 ** (num_of_digits - 1))
        num_of_digits -= 1
    return total
num = -1234
print(f"Sum of all digits of {num}: {digits_sum2(num)}")

Sum of all digits of -1234: 10


# 36. Write a program to find the product of all digits of a number n.

In [61]:
def product_digits(n: int) -> int:
    n = abs(n)

    num_of_digits = len(str(n))
    product = 1
    while num_of_digits > 0:
        digit = n // (10 ** (num_of_digits - 1))
        product *= digit
        n %= (10 ** (num_of_digits - 1))
        num_of_digits -= 1
    return product
    
num = -1234
print(f"Product of all digits of {num} = {product_digits(num)}")

Product of all digits of -1234 = 24


In [62]:
def product_digits2(n: int) -> int:
    n = abs(n)
    product = 1
    while n > 0:
        digit = n % 10
        product *= digit
        n //= 10
    return product

num = -1234
print(f"Product of all digits of {num} = {product_digits2(num)}")

Product of all digits of -1234 = 24


# 37. Write a program to reverse a number n.

In [67]:
def reverse_num(n: int) -> int:
    n = abs(n)
    num = ''.join(reversed(str(n)))
    return int(num)
    
num = int(input("Enter the number : "))
print(f"Reversed of number {num} = {reverse_num(num)}")

Enter the number :  123


Reversed of number 123 = 321


In [76]:
def reverse_num2(n: int) -> int:
    n = abs(n)
    num_str = str(n)
    result = ''
    for idx in range(len(num_str)  - 1, -1, -1):
        result += num_str[idx]
    return result

num = -1234
print(f"Reversed of number {num} = {reverse_num2(num)}")

Reversed of number -1234 = 4321


In [82]:
def reverse_num3(n: int) -> int: 
    n = abs(n)
    # using type casting
    num_of_digits = len(str(n))
    reversed_num = 0
    while num_of_digits > 0:
        last_digit = n % 10
        reversed_num = reversed_num * 10 + last_digit
        num_of_digits -= 1
        n //= 10
    return reversed_num

num = -1234
print(f"Reversed of number {num} = {reverse_num3(num)}")

Reversed of number -1234 = 4321


In [83]:
def reverse_num4(n: int) -> int:
    n = abs(n)
    rev_num = 0
    while n > 0:
        last_digit = n % 10
        rev_num = rev_num * 10 + last_digit
        n //= 10
    return rev_num

num = -12345
print(f"Reversed of number {num} = {reverse_num4(num)}")

Reversed of number -12345 = 54321


# 38. Write a program to find the largest digit in a number n.

In [86]:
def largest_digit(n: int) -> int:
    n = abs(n)
    max_digit = 0
    while n > 0:
        digit = n % 10 
        if digit > max_digit:
            max_digit = digit
        n //= 10
    return max_digit
num = -192347
print(f"Largest digit in {num} = {largest_digit(num)}")

Largest digit in -192347 = 9


# 39. Write a program to find the smallest digit in a number n.

In [87]:
def smallest_digit(n: int) -> int:
    n = abs(n)
    min_digit = 0
    while n > 0:
        digit = n % 10
        if digit < min_digit:
            min_digit = digit
        n //= 10
    return  min_digit
num = -1230895
print(f"Smallest digit in {num} = {smallest_digit(num)}")

Smallest digit in -1230895 = 0


# 40. Write a program to count the number of even digits and odd digits in a number n.

In [94]:
def count_even_odd_digits(n: int) -> tuple:
    n = abs(n)
    even_count, odd_count = 0, 0
    while n > 0:
        digit = n % 10
        if digit % 2 == 0:
            even_count += 1
        else:
            odd_count += 1
        n //= 10
    return (even_count, odd_count)
num = -123456789
counts = count_even_odd_digits(num)
print(f"Count of even digits is {counts[0]} and odd digits is {counts[1]} in number {num}.")

Count of even digits is 4 and odd digits is 5 in number -123456789.


# Day 6: 20/07/26

# 41. Write a program to check whether a number n is a palindrome (reads the same reversed).

In [3]:
def check_palindrome_number(n: int) -> bool:
    temp = n
    rev_num = 0
    while temp > 0:
        digit = temp % 10
        rev_num = rev_num * 10 + digit
        temp //= 10
    return True if rev_num == n else False

num = 1234321
print(f"Is {num} is a Palindrome number? {check_palindrome_number(num)}")

Is 1234321 is a Palindrome number? True


# 42. Write a program to replace all zeros in a number n with the digit 5.

In [10]:
def replace_zero_with_5(n: int) -> int:
    new_num = 0
    num_of_digits = len(str(n))
    
    while num_of_digits > 0:
        
        # get the first digit
        digit = n // (10 ** (num_of_digits - 1))
        
        # get the remainder ie. number without the first digit
        n %= (10 ** (num_of_digits - 1))

        # replace digit 0 with digit 5
        if digit == 0:
            digit = 5
            
        new_num = new_num * 10 + digit
        
        num_of_digits -= 1
    return new_num

num = 2040608
print(f"Earlier number = {num}")
print(f"After replacing new number = {replace_zero_with_5(num)}")

Earlier number = 2040608
After replacing new number = 2545658


# 43. Write a program to find the sum of the first and last digit of a number n.

In [11]:
def sum_of_first_and_last_digit_of_num(n: int) -> int:

    # take first digit
    num_of_digits = len(str(n))
    first_digit = n // (10 ** (num_of_digits - 1))
    last_digit = n % 10
    return first_digit + last_digit

num = 123456
print(f"Sum of the first and last digit = {sum_of_first_and_last_digit_of_num(num)}")

Sum of the first and last digit = 7


# Day 7: 22/07/26

# 44. Write a program to read a number and check whether it is prime or not.


In [8]:
def check_prime_number(n: int) -> bool:
    if num >= 2:
        count = 2
        prime_flag = True
        while count < num:
            if num % count == 0:
                prime_flag = False
                return prime_flag
            count += 1
        return prime_flag
    else:
        print('Please, Enter an integer number > 1')

num = int(input("Enter the number : "))
print(f"Is {num} a prime number ? {check_prime_number(num)}")


Enter the number :  105


2 C
3 C
Is 105 a prime number ? False


In [13]:
def check_prime_number2(num: int) -> bool:
    if num >= 2:
        count = 2
        prime_flag = True
        while count * count < num: # time complexity reduced from O(n) -> O(sq.root(n))
            if num % count == 0:
                prime_flag = False
                return prime_flag 
            count += 1
        return prime_flag
    else:
        print("Please, Enter an integer number > 1")

num = int(input("Enter the integer number: "))
print(f"Is {num} a prime number ? {check_prime_number2(num)}")

Enter the integer number:  23


Is 23 a prime number ? True


In [16]:
def check_prime_number3(num: int) -> bool:
    if num <= 1: # handle numbers less than or equal to 1
        print("Enter a valid integer > 1")

    if num <= 3: # 2 and 3 are prime numbers
        return True

    # Eliminate even numbers and multiples of 3
    if num % 2 == 0 or num % 3 == 0:
        return False

    count = 5
    while count * count <= num:
        if num % count == 0 or num % (count + 2) == 0:
            return False
        count += 6
    return True

num = int(input("Enter the integer number: "))
print(check_prime_number3(num))
        

Enter the integer number:  25


False


# Day 8: 25/07/2026

# 45. Write a program to display all prime numbers from 1 to n.


1. Check if end number of range is greater than 1 else return "Enter valid integer number ( > 1) as end of the range."
2. Take each number in range.
3. perform whether that number is prime or not
4. if that number is prime then print it else continue with the next number until number is <= end number of the range.

In [13]:
def prime_num_btw_1_to_n(n: int) -> int:
    # check if num > 1
    if n > 1:
        num = 2
        print(f"Prime numbers between 1 and {n} are: ")
        while num <= n:
            if num == 2 or num == 3:
                print(num, end=" ")
                num += 1
                continue
                
            if num % 2 == 0 or num % 3 == 0:
                num += 1
                continue
                
            count = 5
            prime_flag = True
            while count * count <= num:
                if num % count == 0 or num % (count + 2) == 0:
                    prime_flag = False
                    break
                count += 6
            if prime_flag:
                print(num, end=" ")
            num += 1   
    else:
        return "Enter valid integer number ( > 1) as end of the range"

num = int(input("Enter the end number for the range 1 to n: "))
prime_num_btw_1_to_n(num)

Enter the end number for the range 1 to n:  50


Prime numbers between 1 and 50 are: 
2 3 5 7 11 13 17 19 23 29 31 37 41 43 47 

# 46. Write a program to display the first n prime numbers.

In [27]:
def first_n_prime(n: int) -> int:
    if n > 0:
        count =  n
        num = 2
        while count > 0:
            if num == 2 or num == 3:
                print(num, end=" ")
                num += 1
                count -= 1
                continue

            if num % 2 == 0 or num % 3 == 0:
                num += 1
                continue

            divisor = 5
            prime_flag = True
            while divisor * divisor <= num:
                if num % divisor == 0 or num % (divisor + 2) == 0:
                    prime_flag = False
                    break
                divisor += 6
            if prime_flag:
                print(num, end=" ")
                count -= 1
            num += 1      
    else:
        print("Enter a natural number.")

num = int(input("Enter the number of prime number you want: "))
print(f"First {num} prime numbers are :")
first_n_prime(num)

Enter the number of prime number you want:  10


First 10 prime numbers are :
2 3 5 7 11 13 17 19 23 29 

# Out of Box interview question

        def myfunction():
            return "Catch me if you can!"
        myfunction()
        print("Done")

The interviewer asks the candidate that if you run the above code snippet it will print Done but if I want to get the returned value of the function how can we get it without modifying the code snippet above.

In [2]:
import sys

# define the tracing function to intercept the returned value
def trace_return(frame, event, arg):
    if event == "return" and frame.f_code.co_name == "myfunction":
        print(f'Intercepted value is "{arg}"')
    return trace_return

# Enable the tracer
sys.settrace(trace_return)

# original code snippet
def myfunction():
    return "Catch me if you can!"
myfunction()
print("Done")

# Disable the tracer
sys.settrace(None)

Intercepted value is "Catch me if you can!"
Done


# Day 9: 28/7/2026

# 47. Write a program to check whether a number is an Armstrong number.

Procedure:

1. get number of digits in number 'N' -> power
2. iterate over digits( dg1, dg2, dg3, ...) of number 'N':
3. create a new number 'New_N'= dg1^power + dg2^power + dg3^power + ...
4. return True if 'New_N' == 'N' else False for proving a number as Armstrong number.
   

In [4]:
def check_armstrong_number(n: int) -> bool:
    n = abs(n)
    num_of_digits = len(str(n))
    new_num = 0
    temp = n
    while temp > 0:
        digit = temp % 10
        new_num += digit ** num_of_digits
        temp //= 10
    if new_num == n:
        return True
    else:
        return False

num1 = 153
print(f"Is {num1} an armstrong number ? {check_armstrong_number(num1)}")
num2 = 123
print(f"Is {num2} an armstrong number ? {check_armstrong_number(num2)}")
        

Is 153 an armstrong number ? True
Is 123 an armstrong number ? False


# Day 10: 06/08/2026

# 48. Write a program to display all Armstrong numbers from 1 to n.

Procedure:

1. Take the number 'N' for deciding the end of the range ie. 1 to N
2. Take each number 'n' from the range 1 to 'N' and check whether that number 'n' is an armstrong number or not.


In [3]:
def all_armstrong_num_in_given_range(n: int) -> int:
    n = abs(n)
    for i in range(1, n + 1):
        num_of_digits = len(str(i))
        new_num = 0
        temp = i
        while temp > 0:
            digit = temp % 10
            new_num += digit ** num_of_digits
            temp //= 10
        if new_num == i:
            print(i, end=" ")

num = 2000
print(f"All armstrong number in the range of 1 to {num} are: ")
all_armstrong_num_in_given_range(num)

All armstrong number in the range of 1 to 2000 are: 
1 2 3 4 5 6 7 8 9 153 370 371 407 1634 

In [6]:
# solving using list comprehension
n = int(input("Enter the upper limit (n): "))

armstrong_numbers = [
   num for num in range(1, n+1) if sum(int(digit) ** len(str(num)) for digit in str(num)) == num
]

print(f"All armstrong numbers in the range of 1 to {n} are: {armstrong_numbers}")

Enter the upper limit (n):  2000


All armstrong numbers in the range of 1 to 2000 are: [1, 2, 3, 4, 5, 6, 7, 8, 9, 153, 370, 371, 407, 1634]


# 49. Write a program to check whether a number is a perfect number.

A perfect number is a positive integer that is equal to the sum of its proper factors (all factors excluding the number itself)

In [10]:
n = int(input("Enter the number to check if is it a Perfect number or not?"))
new_num = 0
for num in range(1, n):
    if n % num == 0:
        new_num += num

if new_num == n:
    print(True)
else:
    print(False)

Enter the number to check if is it a Perfect number or not? 496


True


# 50. Write a program to check whether a number is a strong number (sum of factorials of its digits).

145 = 1! + 4! + 5! = 1 + 24 + 120 = 145

In [14]:
n = int(input("Enter the number to check whether that number is a strong number or not?"))

# function to calculate factorial of a number 
def calculate_factorial(n):
    if n < 0:
        return "Factorial of negative number doesn't exists"
    if n == 0 or n == 1:
        return 1
    else:
        return n * calculate_factorial(n-1)
    
new_num = 0

for digit in str(n):
    new_num += calculate_factorial(int(digit))

if new_num == n:
    print(True)
else:
    print(False)

Enter the number to check whether that number is a strong number or not? 40585


True


# Day 11: 07/08/2026

# 51. Write a program to check whether a number is an automorphic number.

5: (5^2 = 25) (ends in 5)

6: (6^2 = 36) (ends in 6)

25: (25^2 = 625) (ends in 25)

76: (76^2 = 5776) (ends in 76)

625: (625^2 = 390625) (ends in 625)

procedure:

1. given number 'N'.
2. square of the given number 'N^2' => 'M'
3. no of the digits in 'N' => 'num_of_digits_in_N'
4. Extract last 'num_of_digits_in_N' from 'M' => 'last_digits_from_M'
5. if 'last_digits_from_M' equal to 'N' then 'N' is automorphic number.

In [15]:
num = int(input("Enter the number to check if it is an Automorphic number or not: "))

num_of_digits_in_num = len(str(num))
squared_num = num ** 2
print(f"Squared of {num}: {squared_num}")
last_digits_of_squared_num = int(str(squared_num)[-num_of_digits_in_num:])
print(f"Is {num} an Automorphic number? ")
if last_digits_of_squared_num == num:
    print(True)
else:
    print(False)

Enter the number to check if it is an Automorphic number or not:  40


Squared of 40: 1600
Is 40 an Automorphic number? 
False


In [10]:
num = int(input("Enter the number to check if it is an Automorphic number or not?"))
print("Automorphic" if str(num ** 2).endswith(str(num)) else "Not Automorphic")

Enter the number to check if it is an Automorphic number or not? 76


Automorphic


In [16]:
def is_automorphic(n):
    
    if  n < 0:
        return False

    square = n ** 2
    num_digits = len(str(n))
    divisor = 10 ** num_digits

    return (square % divisor) == n

print(is_automorphic(5))
print(is_automorphic(6))
print(is_automorphic(7))
print(is_automorphic(25))
print(is_automorphic(76))
print(is_automorphic(625))

True
True
False
True
True
True


# 52. Write a program to check whether a number is a Harshad (Niven) number.

18 is Harshad number because Harshad number is an integer which is completely divisible by sum of its individual digits

sum of individual digits of 18 = 1 + 8 = 9

18 is completely divisible by 9 because remainder is 0.


In [19]:
def is_Harshad(num):
    temp = num
    sum_of_digits_of_num = 0
    while temp > 0:
        digit = temp % 10
        sum_of_digits_of_num += digit
        temp //= 10
    return num % sum_of_digits_of_num == 0

print(is_Harshad(18))
print(is_Harshad(12))
print(is_Harshad(21))
print(is_Harshad(171))
print(is_Harshad(15))

True
True
True
True
False


# 53. Write a program to find all factors (divisors) of a number n.


In [24]:
def all_factors(num):
    factors = []
    for i in range(1, num + 1):
        if num % i == 0:
            factors.append(i)
            
    return factors

print(all_factors(15))
print(all_factors(100))
print(all_factors(24))
print(all_factors(36))

[1, 3, 5, 15]
[1, 2, 4, 5, 10, 20, 25, 50, 100]
[1, 2, 3, 4, 6, 8, 12, 24]
[1, 2, 3, 4, 6, 9, 12, 18, 36]


# 54. Write a program to count the number of factors of a number n.


In [23]:
def num_of_factors(num):
    factors = []
    for i in range(1, num+1):
        if num % i == 0:
            factors.append(i)

    return len(factors)

print(num_of_factors(15))
print(num_of_factors(100))

4
9


# 55. Write a program to find the GCD (HCF) of two numbers.

In [26]:
def HCF(num1, num2):
    
    factors_num1 = []
    factors_num2 = []
    
    for i in range(1, num1 + 1):
        if num1 % i == 0:
            factors_num1.append(i)
    
    for j in range(1, num2 + 1):
        if num2 % j == 0:
            factors_num2.append(j)

    intersection = list(set(factors_num1) & set(factors_num2))
    return intersection[-1]

print(HCF(24, 36))
print(HCF(3, 5))

12
1


In [37]:
def hcf_euclidean(num1, num2):
    while num2 != 0:
        num1, num2 = num2, num1 % num2
    return num1

print(hcf_euclidean(24, 36))

12


# 56. Write a program to find the LCM of two numbers

In [38]:
def lcm(num1, num2):
    temp_num1, temp_num2 = max(num1, num2), min(num1, num2)
    
    while temp_num2 != 0:
        temp_num1, temp_num2 = temp_num2, temp_num1 % temp_num2
    hcf = temp_num1
    lcm = (num1 * num2) // hcf
    return lcm

print(lcm(2,3))

6


# Day 12: 12/08/2026

# 57. Write a program to display the first n terms of the Fibonacci series.

F0 = 0, F1 = 1, F2 = 1, F3 = 2, F4 = 3, F5 = 5, F6 = 8, F7 = 13, F8 = 21, F9 = 34

Procedure:

1. Take a number 'n' from user to get the number of Fibonacci terms to print.
2. Initialize a counter 'C' that keeps track of number of elements got printed ie. C = 1
3. if n > 0 then follow next Steps else inform the user to enter value greater than 0.
4. print current_term = previous_term + previous_previous_term & increase C by 1
5. Eg. if n = 1 => print first term F0 = 0 and increase C by 1 ie. C = 1
6. Eg. if n >= 2 => print 2 terms F0 = 0, F1 = 1 & set C = 2
7. For n > 2 : then print current_term = previous_term + previous_previous_term & increase C by 1 and make previous_previous_term = previous_term & previous_term = current_terms eg. for 3rd term F2 = F1 + F0 & C += 1 then for 4th term F3 = F2 + F1 & C += 1
8. Stop repeating (Step 7) till counter 'C' becomes equal to 'n'
   

In [15]:
n = int(input("Enter the number of terms of Fibonacci series: "))
num_of_terms = 0
prev_prev_term = 0
prev_term = 1

if n > 0:
    print(f"First {n} terms of Fibonacci series are : ")
    if n == 1:
        print (prev_prev_term, end=" ")
        num_of_terms += 1
    elif n >= 2:
        print(prev_prev_term, prev_term, end=" ")
        num_of_terms = 2
    while n > 2 and num_of_terms < n :
        curr_term = prev_term + prev_prev_term
        prev_prev_term = prev_term
        prev_term = curr_term
        print(curr_term, end=" ")
        num_of_terms += 1
else:
    print("Please enter an integer value greater than 0")


Enter the number of terms of Fibonacci series:  20


First 20 terms of Fibonacci series are : 
0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 610 987 1597 2584 4181 

In [29]:
def Fibonacci_series(n):
    if n <= 0 or type(n) != int:
        print("Please enter an integer value greater than 0.")
    else:
        num_of_terms = 0
        if n == 1:
            print(0)
            num_of_terms = 1
        if n >= 2:
            prev_prev_term = 0
            prev_term = 1
            print(prev_prev_term, prev_term, end=" ")
            num_of_terms = 2
        while n > 2 and num_of_terms < n:
            curr_term = prev_term + prev_prev_term
            prev_prev_term = prev_term
            prev_term = curr_term
            num_of_terms += 1
            print(curr_term, end=" ")

Fibonacci_series(-2)
Fibonacci_series(0)
Fibonacci_series(5.6)
Fibonacci_series(1)
Fibonacci_series(2)
print()
Fibonacci_series(3)
print()
Fibonacci_series(20)

Please enter an integer value greater than 0.
Please enter an integer value greater than 0.
Please enter an integer value greater than 0.
0
0 1 
0 1 1 
0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 610 987 1597 2584 4181 

In [41]:
def recursive_fibonacci_series(n):
    if n <= 0:
        return 0
    elif n == 1:
        return 1
    else:
        return recursive_fibonacci_series(n-1) + recursive_fibonacci_series(n-2)

for i in range(20):
    print(recursive_fibonacci_series(i), end=" ")


0 1 1 2 3 5 8 13 21 34 55 89 144 233 377 610 987 1597 2584 4181 

# 58. Write a program to find the sum of the first n terms of the Fibonacci series.

In [45]:
def recursive_fibo(n):
    if n <= 0:
        return 0
    elif n == 1:
        return 1
    else: 
        return recursive_fibo(n-1) + recursive_fibo(n-2)
fibo_sum = 0
n = 4
print(f"Sum of first {n} terms of Fibonacci series is :")
for i in range(n):
    fibo = recursive_fibo(i)
    fibo_sum += fibo
print(fibo_sum)
    

Sum of first 4 terms of Fibonacci series is :
4


# 59. Write a program to find the sum of the series 1 + 2 + 3 + ... + n.

In [47]:
N = int(input("Enter the number 'N': "))
print(f"Sum of the series 1 + 2 + 3 + ... + {N} : {(N *(N+1))/2}")

Enter the number 'N':  10


Sum of the series 1 + 2 + 3 + ... + 10 : 55.0


# 60. Write a program to find the sum of the series 1^2 + 2^2 + 3^2 + ... + n^2.

In [51]:
N = int(input("Enter the number 'N': "))
print(f"Sum of the series 1^2 + 2^2 + 3^2 + ... + {N}^2 = {(N * (N+1)*(2*N + 1))/6}")

Enter the number 'N':  10


Sum of the series 1^2 + 2^2 + 3^2 + ... + 10^2 = 385.0


# 61. Write a program to find the sum of the series 1 + 1/2 + 1/3 + ... + 1/n.


In [1]:
n = int(input("Enter the number: "))
summation = 0
for i in range(1, n+1):
    summation += 1/i
print(summation)

Enter the number:  5


2.283333333333333


# 62. Write a program to find the value of x raised to the power y without using inbuilt power.


In [2]:
def xraisetoy(x,y):
    # if power is 0
    if y == 0:
        return 1
    # if power is negative
    is_negative = y < 0

    # if power is negative make it positive 
    y = abs(y)

    result = 1
    current_product = x
    while y > 0:
        if y % 2 == 1:
            result *= current_product

        current_product *= current_product
        y //= 2

    return 1 / result if is_negative else result

xraisetoy(2,-4)
    

0.0625

# Day 13: 1/9/2026

# 63. Write a program to print a right-angled triangle pattern of stars of height n.


In [2]:
# n = 5

# *
# * *
# * * *
# * * * *
# * * * * *

n = int(input("Enter the height of right-angled triangle: "))
for row in range(1, n + 1):
    for col in range(1, row + 1):
        print('*', end=" ")
    print()

Enter the height of right-angled triangle:  10


* 
* * 
* * * 
* * * * 
* * * * * 
* * * * * * 
* * * * * * * 
* * * * * * * * 
* * * * * * * * * 
* * * * * * * * * * 


In [9]:
n = int(input("Enter the height of right-angle triangle: "))

for row in range(1, n + 1):
    print("* " * row)

Enter the height of right-angle triangle:  10


* 
* * 
* * * 
* * * * 
* * * * * 
* * * * * * 
* * * * * * * 
* * * * * * * * 
* * * * * * * * * 
* * * * * * * * * * 


# 64. Write a program to print an inverted right-angled triangle pattern of stars of height n.


In [7]:
# n = 5

# * * * * *
# * * * *
# * * *
# * *
# *

n = int(input("Enter the height of inverted right-angled triangle pattern: "))

for row in range(1, n + 1):
    for col in range(n + 1 - row , 0, -1):
        print('*', end=" ")
    print()

Enter the height of inverted right-angled triangle pattern:  10


* * * * * * * * * * 
* * * * * * * * * 
* * * * * * * * 
* * * * * * * 
* * * * * * 
* * * * * 
* * * * 
* * * 
* * 
* 


In [10]:
n = int(input('Enter the height of inverted right-angle triangle: '))

for col in range(n, 0, -1):
    print('* ' * col)

Enter the height of inverted right-angle triangle:  5


* * * * * 
* * * * 
* * * 
* * 
* 


# 65. Write a program to print a pyramid pattern of stars of height n.


In [25]:
# height = 6

#      *
#     ***
#    *****
#   *******
#  *********
# ***********

height = int(input('Enter the height of pyramid: '))
for row in range(1, height + 1):
    # inner for loop for white spaces
    for space in range(height - row):
        print(' ', end=' ')

    # inner for loop for printing '*'
    for star in range(2 * row -1):
        print('*', end=' ')
    print()


Enter the height of pyramid:  6


          * 
        * * * 
      * * * * * 
    * * * * * * * 
  * * * * * * * * * 
* * * * * * * * * * * 


In [33]:
height = int(input("Enter the height: "))

for row in range(1, height + 1):
    print(" " * (height - row) + "*" * (2 * row -1))

Enter the height:  6


     *
    ***
   *****
  *******
 *********
***********


In [18]:
h = int(input('Enter the height: '))

for row in range(h-1, 0, -1):
    print('#' * row)

Enter the height:  6


#####
####
###
##
#


In [14]:
h = int(input('Enter the height: '))
for row in range(1, h + 1):
    for col in range(1, h + 1 + row - 1):
        print('*', end=' ')
    print()

Enter the height:  6


* * * * * * 
* * * * * * * 
* * * * * * * * 
* * * * * * * * * 
* * * * * * * * * * 
* * * * * * * * * * * 


# 66. Write a program to print a number triangle (row i contains numbers 1 to i).


In [11]:
n = int(input("Enter the height of number right-angle triangle: "))

for row in range(1, n + 1):
    for col in range(1, row + 1):
        print(col, end=' ')
    print()

Enter the height of number right-angle triangle:  5


1 
1 2 
1 2 3 
1 2 3 4 
1 2 3 4 5 


# 67. Write a program to print Pascal's triangle for n rows.

    1
    1 1
    1 2 1
    1 3 3 1
    1 4 6 4 1
    1 5 10 10 5 1
    1 6 15 20 15 6 1
    ...

Steps:

    1. 1st row of pascal triangle contain 1.
    2. from 2nd row onwards 
    2.1 each row start with 1 and end with 1.
    2.2 Calculate next element till current row 2nd last element: curr_row_next_element[postion] =  prev_row[postion - 1] + prev_row[postion]
    

In [43]:
def generate_pascal_triangle(n):

    # initialize the pascal triangle with the first row having 1 as single element
    pascal_triangle = [[1]]

    for row in range(1, n):
        
        # initialize the curr_row with 1
        curr_row = [1]

        prev_row = pascal_triangle[row-1]
        for position in range(1, len(prev_row)):
            curr_row_next_element = prev_row[position - 1] + prev_row[position]
            curr_row.append(curr_row_next_element)

        # end the curr_row with 1
        curr_row.append(1)
        pascal_triangle.append(curr_row)

    return pascal_triangle

generate_pascal_triangle(7)

[[1],
 [1, 1],
 [1, 2, 1],
 [1, 3, 3, 1],
 [1, 4, 6, 4, 1],
 [1, 5, 10, 10, 5, 1],
 [1, 6, 15, 20, 15, 6, 1]]

In [51]:
import math

def pascal_triangle_combination(n):
    return [[math.comb(row, col) for col in range(row + 1)] for row in range(n)]

print(pascal_triangle_combination(5))

[[1], [1, 1], [1, 2, 1], [1, 3, 3, 1], [1, 4, 6, 4, 1]]


In [53]:
def getRow(rowIndex: int) -> list[int]:
        pascal_triangle = [[math.comb(row, col) for col in range(row + 1)] for row in range(rowIndex+1)]
        return pascal_triangle[rowIndex]

getRow(3)

[1, 3, 3, 1]

# Day 14: 21/9/2026

# 68. Write a program to find the length of a string without using an inbuilt function.


In [4]:
def stringLength(s: str) -> int:
    length = 0
    for char in s:
        length += 1
    return length

s = 'arvind good'
print(stringLength(s))
    

11


# 69. Write a program to count the number of vowels and consonants in a string.


1. initialize the vowels_counter and consonants_counter with the value of 0.
2. convert the given string to a proper lower case string using ''.join(str.split(' ')).lower() function
3. create a list of vowels
4. run a for loop for the given length of the string
   
   4.1  check if each character in the given string belongs to vowels_list if yes then increase the vowel counter by 1 else consonant counter by 1

5. return a list [vowels_counter, consonants_counter]

In [10]:
def countVowelsAndConsonants(s: str):
    vowels_counter, consonants_counter = 0, 0
    proper_lower_case_input_string = ''.join(s.split(' ')).lower()
    vowels_list = ['a', 'e', 'i', 'o', 'u']

    for char in proper_lower_case_input_string:
        if char in vowels_list:
            vowels_counter += 1
        else:
            consonants_counter += 1

    return [vowels_counter, consonants_counter]

s = "The quick brown fox jumps over the lazy dog"
print(countVowelsAndConsonants(s))
    

[11, 24]


# 70. Write a program to count the number of words in a sentence.

1. break a sentence into words using str.split(' ').
2. create a list to store the words of a sentence by checking if each word is consists of alphabets only.
3. calculate length of list which gives counts of words in a given sentence.

In [1]:
def countwords(sentence):
    words = [word for word in sentence.split(' ')] # for simple sentences
    return len(words)
s = 'hello, world'
print(countwords(s))

2


In [4]:
import re
def countwords2(sentence):
    words = re.findall(r'\b\w+\b', sentence)
    return len(words)

s = "Hello, world! Python is great, isn't it?"
print(countwords2(s))

8


# 71. Write a program to reverse a string.

1. Read each character of the string from backwards.
2. Concatenate the read character to an empty string.


In [9]:
def reverseString(s):
    reversestr = ""
    for idx in range(len(s)-1, -1, -1):
        reversestr += s[idx]

    return reversestr

s = "Hello, world! Arvind is best and can learn anything fast."
print(reverseString(s))

.tsaf gnihtyna nrael nac dna tseb si dnivrA !dlrow ,olleH


In [10]:
def reverseString2(s):
    return s[::-1]

s = "Hello, world! Arvind is best and can learn anything fast."
print(reverseString2(s))

.tsaf gnihtyna nrael nac dna tseb si dnivrA !dlrow ,olleH


In [11]:
def reverseString3(s):
    reversestr = "".join(reversed(s))
    return reversestr

s = "Hello, world! Arvind is best and can learn anything fast."
print(reverseString3(s))

.tsaf gnihtyna nrael nac dna tseb si dnivrA !dlrow ,olleH


# 72. Write a program to check whether a string is a palindrome.

1. First, convert the given string to lowercase and check whether the string is alphanumeric (using the str.isalnum() function).
2. Reverse the converted string and assign it to a variable.
3. Check whether the reversed string is the same as the converted string, then return True or False.

In [14]:
def checkPalindromeString(s):
    s = "".join(char.lower() for char in s if char.isalnum()) 
    reversestr = ""
    for idx in range(len(s)-1, -1, -1):
        reversestr += s[idx]

    return reversestr == s

s = 'aabbaa'
print(checkPalindromeString(s))
s2 = "A man, a plan, a canal: Panama"
print(checkPalindromeString(s2))

True
True


# 73. Write a program to convert a string to uppercase and lowercase without inbuilt case functions.
1. initialize an empty output string.

2.   if case is UPPER, then convert the string to uppercase. [A-Z => 65 - 90]
    iterate through the string

        if the char is alphabet

           if lowercase (ie. ASCII value in between 97 to 122) then subtract 32 from its ASCII value to get its corresponding Uppercase ASCII value then add that uppercase letter to the output string.

           else simply add the char to the output string.

        if the char is not alphabet then simply add it to the output string.

4. else if case is LOWER then convert the string to lowercase. [a-z => 97 - 122]

   iterate throught the string

       if the char is alphabet

           if uppercase (ie. ASCII value in between 65 to 90) then add 32 to its ASCII value to get its corresponding Lowercase ASCII value then add that lowercase letter to the the output string.

           else simply add the char to the output string.

       if the char is not alphabet then simply add it to the output string.

       

In [3]:
def changecase(text, case):
    output_string = ""
    if case == "UPPER": # for converting text to UPPER CASE
        for char in text:
            if char.isalpha():
                if char.islower():
                    new_char = chr(ord(char) - 32)
                    output_string += new_char
                else:
                    output_string += char
            else:
                output_string += char
    elif case == "lower":
        for char in text:
            if char.isalpha():
                if char.isupper():
                    new_char = chr(ord(char) + 32)
                    output_string += new_char
                else:
                    output_string += char
            else:
                output_string += char
        
    return output_string

s = 'arVinD iS a GooD peRSon! Who is goNNa be ricH.'
case = "UPPER"
print(changecase(s, case))
s2 = 'arVinD iS a GooD peRSon! Who is goNNa be ricH.'
case2 = 'lower'
print(changecase(s2, case2))

ARVIND IS A GOOD PERSON! WHO IS GONNA BE RICH.
arvind is a good person! who is gonna be rich.


# 74. Write a program to count the frequency of each character in a string.


In [5]:
def charFrequency(text):
    freq_map = dict()
    for char in text:
        if char in freq_map:
            freq_map[char] += 1
        else:
            freq_map[char] = 1
    return freq_map

s = "hello!, world."
print(charFrequency(s))
s2 = 'arVinD iS a GooD peRSon! Who is goNNa be ricH.'
print(charFrequency(s2))

{'h': 1, 'e': 1, 'l': 3, 'o': 2, '!': 1, ',': 1, ' ': 1, 'w': 1, 'r': 1, 'd': 1, '.': 1}
{'a': 3, 'r': 2, 'V': 1, 'i': 4, 'n': 2, 'D': 2, ' ': 9, 'S': 2, 'G': 1, 'o': 5, 'p': 1, 'e': 2, 'R': 1, '!': 1, 'W': 1, 'h': 1, 's': 1, 'g': 1, 'N': 2, 'b': 1, 'c': 1, 'H': 1, '.': 1}


In [6]:
def charFrequency2(text):
    freq_map = {}

    for char in text:
        freq_map[char] = freq_map.get(char, 0) + 1
    return freq_map

s = "hello!, world."
print(charFrequency2(s))
s2 = 'arVinD iS a GooD peRSon! Who is goNNa be ricH.'
print(charFrequency2(s2))

{'h': 1, 'e': 1, 'l': 3, 'o': 2, '!': 1, ',': 1, ' ': 1, 'w': 1, 'r': 1, 'd': 1, '.': 1}
{'a': 3, 'r': 2, 'V': 1, 'i': 4, 'n': 2, 'D': 2, ' ': 9, 'S': 2, 'G': 1, 'o': 5, 'p': 1, 'e': 2, 'R': 1, '!': 1, 'W': 1, 'h': 1, 's': 1, 'g': 1, 'N': 2, 'b': 1, 'c': 1, 'H': 1, '.': 1}


In [7]:
from collections import Counter

text = 'hello, world'

frequencies = Counter(text)

print(dict(frequencies))

{'h': 1, 'e': 1, 'l': 3, 'o': 2, ',': 1, ' ': 1, 'w': 1, 'r': 1, 'd': 1}


# Date: 01/10/2026

# 75. Write a program to remove all spaces from a string.


In [10]:
def removespaces(s):
    removedspacesstring = s.split(' ')
    return ''.join(word for word in removedspacesstring)

s = 'Arvind say     Hello,  world.'
print(removespaces(s))

ArvindsayHello,world.


In [8]:
def removeallspaces(user_text):
    text_without_spaces = ""
    for char in user_text:
        if char != ' ':
            text_without_spaces += char
    return text_without_spaces

s = 'Arvind say Hello, world.'
print(removeallspaces(s))


ArvindsayHello,world.


# 76. Write a program to check whether two strings are anagrams of each other.

In [13]:
def anagramscheck(string1, string2):
    # remove all spaces from both strings
    string1_without_spaces = ''
    for char in string1:
        if char != ' ':
            string1_without_spaces += char

    string2_without_spaces = ''
    for char in string2:
        if char != ' ':
            string2_without_spaces += char

    # convert both strings to lowercase
    string1_without_spaces_lowercase = string1_without_spaces.lower()
    string2_without_spaces_lowercase = string2_without_spaces.lower()

    # Conditions of both the strings to be anagrams of each other are 1. Same lenght 2. Identical characters 3. same character frequencies.

    # check if length of both the strings are same or not.
    if len(string1_without_spaces_lowercase) != len(string2_without_spaces_lowercase):
        return False

    freq_map_string1 = {}
    for char in string1_without_spaces_lowercase:
        freq_map_string1[char] = freq_map_string1.get(char, 0) + 1

    freq_map_string2 = {}
    for char in string2_without_spaces_lowercase:
        freq_map_string2[char] = freq_map_string2.get(char, 0) + 1
        
    # check if identical characters are their in both the strings or not.
    if freq_map_string1.keys() != freq_map_string2.keys():
        return False
    # check if same character frequencies are their or not.
    else:
        for char in freq_map_string1.keys():
            if freq_map_string1.get(char) != freq_map_string2.get(char):
                return False
                
    return True

s1, s2 = 'Listen', 'Silent'
print(anagramscheck(s1, s2))

s3, s4 = "Slot machines", "Cash lost in me"
print(anagramscheck(s3, s4))
s5, s6 = "look", "Lock"
print(anagramscheck(s5, s6))
s7, s8 = "Apple", "pale"
print(anagramscheck(s7, s8))

True
True
False
False


In [14]:
# using str.replace() function to replace white spaces with nothing and sorted() function to get the sorted list of characters of both the strings.
def checkanagrams(text1, text2):
    text1 = text1.replace(" ", "").lower()
    text2 = text2.replace(" ", "").lower()

    if len(text1) != len(text2):
        return False

    return sorted(text1) == sorted(text2)

s1, s2 = 'Listen', 'Silent'
print(checkanagrams(s1, s2))

s3, s4 = "Slot machines", "Cash lost in me"
print(checkanagrams(s3, s4))
s5, s6 = "look", "Lock"
print(checkanagrams(s5, s6))
s7, s8 = "Apple", "pale"
print(checkanagrams(s7, s8))

True
True
False
False


# Date: 02/10/2026

# 77. Write a program to find the first non-repeating character in a string.

1. First, create a frequency_map of chars in a given string.
2. Then iterate through the frequency_map dictionary

        2.1 Check if the frequency of the char is equal to 1; then return that char
3. If all the chars repeat more than once, then return None

In [6]:
def firstNonRepeatingChar(string):
    freq_map = {}
    for char in string:
        freq_map[char] = freq_map.get(char, 0) + 1

    for char in freq_map:
        if freq_map[char] == 1:
            return char

    # if none of the character appears once
    return None

text = "swiss"
print(firstNonRepeatingChar(text))

w


In [7]:
# using Counter from collections
from collections import Counter

def firstNonRepeatingChar2(string):
    counts = Counter(string)

    for char in string:
        if counts[char] == 1:
            return char

    # if none of the characters appears once
    return None

text = "swiss"
print(firstNonRepeatingChar2(text))

w


# 78. Write a program to replace all occurrences of one character with another character in a string.

1. This can be done using str.replace("char_to_replaced", "char_to_be_replaced_with").

In [11]:
def replaceAllOccurrencesOfAChar(string, char_to_be_replace, char_to_be_replaced_with):
    if char_to_be_replace in string:
        return string.replace(char_to_be_replace, char_to_be_replaced_with)
    else:
        return f"{char_to_be_replace} is not in the given string: '{string}'"

text = 'Arvind is following the norms of Winter Arc.'
c1, c2 = 'z', 'O'
print(replaceAllOccurrencesOfAChar(text, c1, c2))

z is not in the given string: 'Arvind is following the norms of Winter Arc.'


1. All occurrences of a character can be replaced with another character in a string using brute force of iterating through the entire string.

In [14]:
def replaceAllOccurrencesOfACharWithAnotherChar(string, char_to_be_replace, char_to_be_replaced_with):
    output_string = ""
    if char_to_be_replace not in string:
        return f"{char_to_be_replace} is not in the given string: '{string}'"
    for char in string:
        if char == char_to_be_replace:
            output_string += char_to_be_replaced_with
        else:
            output_string += char
    return output_string

text = 'Arvind is following the norms of Winter Arc.'
c1, c2 = 'i', 'I'
print(replaceAllOccurrencesOfACharWithAnotherChar(text, c1, c2))

ArvInd Is followIng the norms of WInter Arc.


# 79. Write a program to toggle the case of each character in a string.

A-Z = 65-90

a-z = 97-122

1. Initialise an output string.
2. Iterate through the given string char by char.
   
        2.1  Check if char is alphabet then
            2.1.1 Check if char is lower then add 32 to the ascii value of lowercase alphabet then append the corresponding uppercase char to the output string.
            else: subtract 32 to the ascii value of uppercase alphabet then append the corresponding lowercase char to the output string.
        else: append the char to the output string.

In [5]:
def togglecase(string):
    output_string = ""
    for char in string:
        if char.isalpha():
            if char.islower():
                output_string += chr(ord(char) - 32)
            else:
                output_string += chr(ord(char) + 32)
        else:
            output_string += char
    return output_string

text = 'HeLLo, WoRlD! My sElF ArvInD AND I hAve sTarTed WinTER ARC ChaLLenGE.'
print(togglecase(text))

hEllO, wOrLd! mY SeLf aRViNd and i HaVE StARtED wINter arc cHAllENge.


# Date: 03/10/2026

# Arrays

# 80. Write a program to read n elements into an array and print them.


In [2]:
n = int(input("Enter the length of an array: "))
arr = []
for index in range(n):
    element = int(input(f"Enter the {index + 1} element of the array: "))
    arr.append(element)
print("Elements of the array are: ")
for element in arr: 
    print(element)
print(f"Array: {arr}")

Enter the length of an array:  5
Enter the 1 element of the array:  10
Enter the 2 element of the array:  20
Enter the 3 element of the array:  30
Enter the 4 element of the array:  40
Enter the 5 element of the array:  50


Elements of the array are: 
10
20
30
40
50
Array: [10, 20, 30, 40, 50]


# 81. Write a program to find the sum and average of all elements in an array.


In [7]:
def sumAndAverage(arr):
    total = sum(arr)
    avg = total / len(arr)
    return total, avg

arr = [11,23,36,50,65]
total, avg = sumAndAverage(arr)
print(f"Sum and average of all elements in an array {arr} is {total} and {avg} respectively.")

Sum and average of all elements in an array [11, 23, 36, 50, 65] is 185 and 37.0 respectively.


# 82. Write a program to find the largest and smallest element in an array.


In [8]:
def maxminelement(arr): # => N + N = 2N operations ==> TC = O(N)
    max_element = max(arr) # => N operations
    min_element = min(arr) # => N operations 
    return max_element, min_element

arr = [24, 4, 86, 1, 99]
max_element, min_element = maxminelement(arr)
print(f"Largets and smallest element in an array {arr} is {max_element} and {min_element} respectively.")

Largets and smallest element in an array [24, 4, 86, 1, 99] is 99 and 1 respectively.


In [9]:
def largestSmallestElement(arr):
    if not arr:
        return None, None
        
    max_element = float('-inf')
    min_element = float('inf')
    for element in arr:
        if element > max_element:
            max_element = element
        if element < min_element:
            min_element = element
    return max_element, min_element

arr = [24, 4, 86, 1, 99]
max_element, min_element = largestSmallestElement(arr)
print(f"Largets and smallest element in an array {arr} is {max_element} and {min_element} respectively.")

Largets and smallest element in an array [24, 4, 86, 1, 99] is 99 and 1 respectively.


# 83. Write a program to count the number of even and odd elements in an array.


In [7]:
def OddEvenElementCount(arr):
    # If the array is empty
    if not arr:
        return {"odd_count": None, "even_count": None, "reason": "Given array is empty." }

    odd_count, even_count = 0, 0
    for num in arr:
        if type(num) is int:
            if num % 2 == 0:
                even_count += 1
            else:
                odd_count += 1
        else:
            return {"odd_count": None, "even_count": None, "reason": "At least one element in the given array is not an integer."}

    return {"odd_count": odd_count, "even_count": even_count, "reason": "All elements of the given array were integers."}


arr1 = [1,2,3,4,5,6,7,8,9,10]
print(OddEvenElementCount(arr1))
arr2 = []
print(OddEvenElementCount(arr2))
arr3 = [1,2,3,4,5,6,7.7,8,9,10]
print(OddEvenElementCount(arr3))
arr4 = [-83, 9, 6, 22, -93, 72, 81, 95, 77, -136, 59]
print(OddEvenElementCount(arr4))

{'odd_count': 5, 'even_count': 5, 'reason': 'All elements of the given array were integers.'}
{'odd_count': None, 'even_count': None, 'reason': 'Given array is empty.'}
{'odd_count': None, 'even_count': None, 'reason': 'At least one element in the given array is not an integer.'}
{'odd_count': 7, 'even_count': 4, 'reason': 'All elements of the given array were integers.'}


In [9]:
# Advance solution
def oddEvenCount2(arr):
    if not arr:
        return None, None
    even_count = sum(1 for num in arr if num % 2 == 0)
    odd_count = len(arr) - even_count
    return even_count, odd_count

arr1 = [1,2,3,4,5,6,7,8,9,10]
print(oddEvenCount2(arr1))
arr2 = []
print(oddEvenCount2(arr2))
arr3 = [1,2,3,4,5,6,7.7,8,9,10]
print(oddEvenCount2(arr3))
arr4 = [-83, 9, 6, 22, -93, 72, 81, 95, 77, -136, 59]
print(oddEvenCount2(arr4))

(5, 5)
(None, None)
(5, 5)
(4, 7)


# 84. Write a program to search for an element in an array (linear search).


In [11]:
def linearSearch(arr, element):
    if not arr:
        return 'Empty Array!'
    for idx in range(len(arr)):
        if arr[idx] == element:
            return idx
    return -1
    
arr = [-83, 9, 6, 22, -93, 72, 81, 95, 77, -136, 59]
# element = 81
element = -81
index = linearSearch(arr, element)
if index == -1:
    print(f"{element} is not present in {arr}.")
else:
    print(f"{element} is present in {arr} at index position {index}.")

-81 is not present in [-83, 9, 6, 22, -93, 72, 81, 95, 77, -136, 59].


# 85. Write a program to reverse the elements of an array.


In [13]:
def reverseArrElements(arr): 
    # Two pointer - Swap approach, Time Complexity = O(N), Space Complexity = O(1)
    for itr in range(len(arr)//2):
        front, rear = itr, len(arr) - 1 - itr
        arr[itr], arr[len(arr) - 1 - itr] = arr[len(arr) - 1 - itr], arr[itr]
    return arr

arr1 = [1,2,3,4,5,6,7]
arr2 = [1,2,3,4,5,6,7,8]
arr3 = []
print(reverseArrElements(arr1))
print(reverseArrElements(arr2))
print(reverseArrElements(arr3))

[7, 6, 5, 4, 3, 2, 1]
[8, 7, 6, 5, 4, 3, 2, 1]
[]


In [19]:
def reverseArrElements1(arr): # Most Optimised approach
    return arr[::-1]

arr1 = [1,2,3,4,5,6,7]
arr2 = [1,2,3,4,5,6,7,8]
arr3 = []
print(reverseArrElements1(arr1))
print(reverseArrElements1(arr2))
print(reverseArrElements1(arr3))

[7, 6, 5, 4, 3, 2, 1]
[8, 7, 6, 5, 4, 3, 2, 1]
[]


In [15]:
def reverseArrElements2(arr): # This approach is the fastest execution approach as it runs in highly optimized C-code, therefore useful in real-world runtime. 
    return list(reversed(arr))

arr1 = [1,2,3,4,5,6,7]
arr2 = [1,2,3,4,5,6,7,8]
arr3 = []
print(reverseArrElements2(arr1))
print(reverseArrElements2(arr2))
print(reverseArrElements2(arr3))


[7, 6, 5, 4, 3, 2, 1]
[8, 7, 6, 5, 4, 3, 2, 1]
[]


In [18]:
def reverseArrElements3(arr):
    reverse_arr = []
    for idx in range(len(arr)):
        reverse_arr.append(arr[len(arr) - 1 - idx])
    return reverse_arr

arr1 = [1,2,3,4,5,6,7]
arr2 = [1,2,3,4,5,6,7,8]
arr3 = []
print(reverseArrElements3(arr1))
print(reverseArrElements3(arr2))
print(reverseArrElements3(arr3))

[7, 6, 5, 4, 3, 2, 1]
[8, 7, 6, 5, 4, 3, 2, 1]
[]


# Date: 04/10/2026

# 86. Write a program to find the second largest element in an array.

Sample inputs:-

1) Standard inputs:

   nums = [4,8,2,7,9]; second_largest = 8 # [unsorted]

   
   nums = [1,2,3,4,5]; second_largest = 4 # [ascendingly sorted]
   
   nums = [6,5,4,3,2,1]; second_largest = 5 # [descendingly sorted]

3) Duplicates :

   nums = [9,9,7,3]; second_largest = 7 # [starting duplicates]

   
   nums = [1,2,3,4,4]; second_largest = 3 # [last duplicates]

   
   nums = [8,1,2,8,5,6,8]; second_largest = 6 # [randomly placed duplicates]

5) Negative numbers:

   nums = [-3,-25,-12,-5,-8]; second_largest = -5 # [All negative]

   
   nums = [-7, -4, 0, -9, 5]; second_largest = 0 # [All mixed] 

7) Edge cases:

   nums = [4,4,4,4,4]; expected_output = None

   
   nums = [7]; expected_output = None # [No 2 elements present]

   
   nums = []; expected_output = None # [array list empty]


In [20]:
def secondLargest(arr): # Not an optimised solution, as I gave too much attention to validating the input
    # Check if arr is empty
    if not arr:
        return None

    # Check if there is only one element in the given arr
    if len(arr) == 1:
        return None

    # Check if all elements of the given arr are equal or not
    if all(arr[i] == arr[i + 1] for i in range(len(arr) - 1)):
        return None

    # Check if the given input array is already sorted in ascending order and each element is unique
    if all(arr[i] < arr[i + 1] for i in range(len(arr) - 1)) and len(arr) == len(set(arr)):
        return arr[-2] # return the second last element as the second largest element

    # Check if the given input array is already sorted in descending order and each element is unique
    if all(arr[i] > arr[i + 1] for i in range(len(arr) - 1)) and len(arr) == len(set(arr)):
        return arr[1] # return the 1st index element as the second largest number

    # initial first and second largest element with negative infinity
    first_largest, second_largest = float('-inf'), float('-inf')
    
    for num in arr:
        if num > first_largest:
            second_largest = first_largest
            first_largest = num
        if num > second_largest and num != first_largest:
            second_largest = num
            
    return second_largest


nums1 = [4,4,4,4,4]
print(secondLargest(nums1))
nums2 = [7]
print(secondLargest(nums2))
nums3 = []
print(secondLargest(nums3))
nums4 = [4,8,2,7,9]
print(secondLargest(nums4))
nums5 = [1,2,3,4,5]
print(secondLargest(nums5))
nums6 = [6,5,4,3,2,1]
print(secondLargest(nums6))
nums7 = [9,9,7,3]
print(secondLargest(nums7))
nums8 = [1,2,3,4,4]
print(secondLargest(nums8))
nums9 = [8,1,2,8,5,6,8]
print(secondLargest(nums9))
nums10 = [-3,-25,-12,-5,-8]
print(secondLargest(nums10))
nums11 = [-7, -4, 0, -9, 5]
print(secondLargest(nums11))

None
None
None
8
4
5
7
3
6
-5
0


In [21]:
def secondLargestNum(nums):

    # Check if nums is empty or not OR length of nums is greater than one
    if not nums or len(nums) < 2:
        return None

    # Initialise first and second largest number as negative infinity
    first_largest = second_largest = float('-inf')

    for num in nums:
        # Case 1: Current number is greater than the largest found so far
        if num > first_largest:
            second_largest = first_largest
            first_largest = num
        # Case 2: Current number is in between first and second largest
        elif num > second_largest and num != first_largest:
            second_largest = num

    return second_largest if second_largest != float('-inf') else None

nums1 = [4,4,4,4,4]
print(secondLargestNum(nums1))
nums2 = [7]
print(secondLargestNum(nums2))
nums3 = []
print(secondLargestNum(nums3))
nums4 = [4,8,2,7,9]
print(secondLargestNum(nums4))
nums5 = [1,2,3,4,5]
print(secondLargestNum(nums5))
nums6 = [6,5,4,3,2,1]
print(secondLargestNum(nums6))
nums7 = [9,9,7,3]
print(secondLargestNum(nums7))
nums8 = [1,2,3,4,4]
print(secondLargestNum(nums8))
nums9 = [8,1,2,8,5,6,8]
print(secondLargestNum(nums9))
nums10 = [-3,-25,-12,-5,-8]
print(secondLargestNum(nums10))
nums11 = [-7, -4, 0, -9, 5]
print(secondLargestNum(nums11))

None
None
None
8
4
5
7
3
6
-5
0


# 87. Program to find the third largest element in a given array

In [24]:
def thirdLargest(nums): # pushed on Leetcode too
    if not nums or len(nums) < 3:
        return None

    first_largest = second_largest = third_largest = float('-inf')

    for num in nums:
        # Case 1: current num is the largest found so far
        if num > first_largest:
            third_largest = second_largest
            second_largest = first_largest
            first_largest = num
        # Case 2: current num is in between first and second largest
        elif num > second_largest and num != first_largest:
            third_largest = second_largest
            second_largest = num
        # Case 3: current num is greater than third largest and smaller than both first and second largest
        elif num > third_largest and num != second_largest and num != third_largest:
            third_largest = num

    return third_largest if third_largest != float('-inf') else None

# 88. Write a program to count the frequency of each element in an array.


In [23]:
def elementfrequency(arr):
    freq_map = dict()

    for element in arr:
        freq_map[element] = freq_map.get(element, 0) + 1

    return freq_map

nums = [1,3,1,2,3,8,6,6,9,8,7,3,4,1,1,4,3,7,9]
print(elementfrequency(nums))

{1: 4, 3: 4, 2: 1, 8: 2, 6: 2, 9: 2, 7: 2, 4: 2}


# 89. Program to find Top K most frequent elements in a given array

In [30]:
def topKFrequent(nums: list[int], k: int) -> list[int]:
    freq_map = dict()
    for num in nums:
        freq_map[num] = freq_map.get(num, 0) + 1

    sorted_freq_map = dict(sorted(freq_map.items(), key=lambda x: x[1], reverse=True))
    return list(sorted_freq_map.keys())[:k]

nums1 = [1,1,1,2,2,3]
k1 = 2
print(topKFrequent(nums1, k1))
nums2 = [1]
k2 = 1
print(topKFrequent(nums2, k2))
nums3 = [1,2,1,2,1,2,3,1,3,2]
k3 = 2
print(topKFrequent(nums3, k3))

[1, 2]
[1]
[1, 2]


In [32]:
def topKFrequent1(nums: list[int], k: int) -> list[int]:
    freq_map = dict()
    for num in nums:
        freq_map[num] = freq_map.get(num, 0) + 1
        
    # create a bucket where index = freq
    buckets = [[] for _ in range(len(nums)+1)]
    for num, count in freq_map.items():
        buckets[count].append(num)

    result = []
    for i in range(len(buckets)-1,0,-1): # we iterate backwards because we want to collect the top K 
# most frequent elements, and buckets are indexed by frequency, so larger frequencies are at the end.
        for num in buckets[i]:
            result.append(num)
            if len(result) == k:
                return result

nums1 = [1,1,1,2,2,3]
k1 = 2
print(topKFrequent1(nums1, k1))
nums2 = [1]
k2 = 1
print(topKFrequent1(nums2, k2))
nums3 = [1,2,1,2,1,2,3,1,3,2]
k3 = 2
print(topKFrequent1(nums3, k3))

[1, 2]
[1]
[1, 2]


# Date: 05/10/2026

# 90. Write a program to remove duplicate elements from an array.

1. Convert the array into set data structure and convert it back to list data structure


In [15]:
def removeDuplicates(arr):
    return arr if len(arr) < 2 else list(set(arr)) # using set() the order of elements is not preserved

nums = [1,2,3,2,5,2,8,2]
print(removeDuplicates(nums))
    

[1, 2, 3, 5, 8]


In [12]:
def removeDuplicates1(arr):
    # Check if arr is empty or has a single element
    if len(arr) < 2:
        return arr
    
    freq_map = {}
    for element in arr:
        freq_map[element] = freq_map.get(element, 0) + 1
    result_arr = []
    for num in freq_map:
        result_arr.append(num)
    return result_arr

nums1 = [1,2,3,2,5,2,8,2]
nums2 = []
nums3 = [42]
nums4 = [7, 7, 7, 7, 7]
nums5 = [2,2]
nums6 = [1, 2, 3, 4, 5]
nums7 = [-3, -3, -1, 0, 0, 2]
nums8 = [-10^9, 10^9, 10^9]
nums9 = [1, 1, 2, 2, 3, 4, 4, 5]
print(removeDuplicates1(nums1))
print(removeDuplicates1(nums2))
print(removeDuplicates1(nums3))
print(removeDuplicates1(nums4))
print(removeDuplicates1(nums5))
print(removeDuplicates1(nums6))
print(removeDuplicates1(nums7))
print(removeDuplicates1(nums8))
print(removeDuplicates1(nums9))

[1, 2, 3, 5, 8]
[]
[42]
[7]
[2]
[1, 2, 3, 4, 5]
[-3, -1, 0, 2]
[-1, 3]
[1, 2, 3, 4, 5]


In [18]:
def removeDuplicatesUnsorted(arr):
    seen = set()
    result = []
    for element in arr:
        if element not in seen:
            seen.add(element)
            result.append(element)
    return result

nums1 = [1,2,3,2,5,2,8,2]
removeDuplicatesUnsorted(nums1)

[1, 2, 3, 5, 8]

In [ ]:
# The interview mainly asks to remove duplicates from a sorted array.
def removeDuplicatesSorted(arr):

    if not arr:
        return 0

    write_idx = 1
    for idx in range(1, len(arr)):
        if arr[idx] != arr[idx - 1]:
            arr[write_idx] = arr[idx]
            write_idx += 1

    return write_idx


            

# 91. Write a program to sort an array in ascending order (bubble sort).

=> Just like bubbles rise in water, similarly the smallest or largest element rise to top of the array depending on ascending order or descending order sorting.

arr = [5,1,4,2], sort arr in ascending order using bubble sort algorithm

Iteration 1:

    1. Compare (5,1), since 5 > 1, swap both => [1,5,4,2]
    2. Compare (5,4), since 5 > 4, swap both => [1,4,5,2]
    3. Compare (5,2, since 5 > 2, swap both => [1,4,2,5]
    Now 5 is at its correct location

Iteration 2:

    1. Compare (1,4), since 1 < 4, no swap => [1,4,2,5]
    2. Compare (4,2), since 4 > 2, swap both => [1,2,4,5]
    Now 4 is at its correct location

Iteration 3:

    1. Compare (1,2), since 1 < 2, no swap => [1,2,4,5]
    Since no swapping occurred the resultant array is sorted ascendingly 

In [19]:
def bubbleSort(arr):

    if not arr: 
        return -1
    if len(arr) == 1:
        return arr # since a one-element array is always sorted.

    for i in range(len(arr)):
        swapped = False

        for j in range(len(arr) - i - 1):
            if arr[j] > arr[j+1]:
                # swap both 
                arr[j], arr[j+1] = arr[j+1], arr[j]
                swapped = True

        if not swapped:
            break
    return arr

nums1 = [1,3,4,2,6,9,5]
print(bubbleSort(nums1))
nums2 = [2,3,4,5,6,7,8,9]
print(bubbleSort(nums2))

[1, 2, 3, 4, 5, 6, 9]
[2, 3, 4, 5, 6, 7, 8, 9]


# Date: 06/10/2026

# 92. Write a program to merge two arrays into one.

1. Case 1: The lengths of the two arrays are the same.

    arr1 = [1,2,3,4];    arr2 = [5,6,7,8];    Expected output: output_arr = [1,2,3,4,5,6,7,8]

2. Case 2: The length of the 1st array is greater than the length of the 2nd array.

    arr1 = [1,2,3,4,5,6];    arr2 = [7,8];    Expected output: output_arr = [1,2,3,4,5,6,7,8]

3. Case 3: The length of the 1st array is smaller than the length of the 2nd array.

    arr1 = [1,2,];    arr2 = [3,4,5,6,7,8];    Expected output: output_arr = [1,2,3,4,5,6,7,8]


1. Initialise an empty output_arr.
2. Iterate through the total of (length(arr1) + length(arr2))
3. Take the elements from arr1 one by one and append those elements to output_arr.
4. Once the elements of arr1 are completed, take the elements from arr2 one by one and append those elements to output_arr.

In [7]:
def mergeTwoArr(arr1, arr2):
    n1, n2 = len(arr1), len(arr2)
    output_arr = []
    for idx in range(n1+n2):
        if idx < n1:
            output_arr.append(arr1[idx])
        else:
            output_arr.append(arr2[idx - n1])
    return output_arr

arr1, arr2 =  [1,2,3,4], [5,6,7,8]
print(mergeTwoArr(arr1, arr2))
arr3, arr4 = [11,22,33,44,55,66], [77,88]
print(mergeTwoArr(arr3, arr4))
arr5, arr6 = [10,20], [30,40,50,60,70,80]
print(mergeTwoArr(arr5, arr6))
print(mergeTwoArr(arr2,arr1))
print(mergeTwoArr(arr4,arr3))
print(mergeTwoArr(arr6, arr5))

[1, 2, 3, 4, 5, 6, 7, 8]
[11, 22, 33, 44, 55, 66, 77, 88]
[10, 20, 30, 40, 50, 60, 70, 80]
[5, 6, 7, 8, 1, 2, 3, 4]
[77, 88, 11, 22, 33, 44, 55, 66]
[30, 40, 50, 60, 70, 80, 10, 20]


In [8]:
# merging using '+' operator
def mergeTwoArr2(arr1, arr2):
    return arr1 + arr2

arr1, arr2 =  [1,2,3,4], [5,6,7,8]
print(mergeTwoArr2(arr1, arr2))
arr3, arr4 = [11,22,33,44,55,66], [77,88]
print(mergeTwoArr2(arr3, arr4))
arr5, arr6 = [10,20], [30,40,50,60,70,80]
print(mergeTwoArr2(arr5, arr6))
print(mergeTwoArr2(arr2,arr1))
print(mergeTwoArr2(arr4,arr3))
print(mergeTwoArr2(arr6, arr5))

[1, 2, 3, 4, 5, 6, 7, 8]
[11, 22, 33, 44, 55, 66, 77, 88]
[10, 20, 30, 40, 50, 60, 70, 80]
[5, 6, 7, 8, 1, 2, 3, 4]
[77, 88, 11, 22, 33, 44, 55, 66]
[30, 40, 50, 60, 70, 80, 10, 20]


In [18]:
# merging using extend() method
def mergeTwoArr3(arr1, arr2):
    arr1.extend(arr2)
    return arr1
arr1, arr2 =  [1,2,3,4], [5,6,7,8]
print(mergeTwoArr3(arr1, arr2))
arr3, arr4 = [11,22,33,44,55,66], [77,88]
print(mergeTwoArr3(arr3, arr4))
arr5, arr6 = [10,20], [30,40,50,60,70,80]
print(mergeTwoArr3(arr5, arr6))
print(mergeTwoArr3(arr2,arr1))
print(mergeTwoArr3(arr4,arr3))
print(mergeTwoArr3(arr6, arr5))

[1, 2, 3, 4, 5, 6, 7, 8]
[11, 22, 33, 44, 55, 66, 77, 88]
[10, 20, 30, 40, 50, 60, 70, 80]
[5, 6, 7, 8, 1, 2, 3, 4, 5, 6, 7, 8]
[77, 88, 11, 22, 33, 44, 55, 66, 77, 88]
[30, 40, 50, 60, 70, 80, 10, 20, 30, 40, 50, 60, 70, 80]


In [19]:
# Merging using Asterisk Unpacking
def mergeTwoArr4(arr1, arr2):
    return [*arr1, *arr2]

arr1, arr2 =  [1,2,3,4], [5,6,7,8]
print(mergeTwoArr4(arr1, arr2))
arr3, arr4 = [11,22,33,44,55,66], [77,88]
print(mergeTwoArr4(arr3, arr4))
arr5, arr6 = [10,20], [30,40,50,60,70,80]
print(mergeTwoArr4(arr5, arr6))
print(mergeTwoArr4(arr2,arr1))
print(mergeTwoArr4(arr4,arr3))
print(mergeTwoArr4(arr6, arr5))


[1, 2, 3, 4, 5, 6, 7, 8]
[11, 22, 33, 44, 55, 66, 77, 88]
[10, 20, 30, 40, 50, 60, 70, 80]
[5, 6, 7, 8, 1, 2, 3, 4]
[77, 88, 11, 22, 33, 44, 55, 66]
[30, 40, 50, 60, 70, 80, 10, 20]


# 93. Write a program to find the sum of all even-indexed and odd-indexed elements separately.

In [20]:
def evenOddIndexedElementSum(arr):
    even_indexed_element_sum = odd_indexed_element_sum = 0
    for idx in range(len(arr)):
        if idx % 2 == 0:
            even_indexed_element_sum += arr[idx]
        else:
            odd_indexed_element_sum += arr[idx]

    return {'even-indexed elements sum': even_indexed_element_sum, 'odd-indexed elements sum': odd_indexed_element_sum}

arr1 = [1, 2, 3, 4, 5, 6, 7, 8]
print(evenOddIndexedElementSum(arr1))


{'even-indexed elements sum': 16, 'odd-indexed elements sum': 20}


# 94. Write a program to check whether a number is prime, using a function/method.


In [23]:
def checkPrimeNumber(number):

    if number <= 1:
        return "Please enter a valid number greater than 1."
    if number <= 3:
        return True
    if number % 2 == 0 or number % 3 == 0:
        return False
    count = 5
    while count * count <= number:
        if (number % count) == 0 or (number % (count + 2) == 0):
            return False
        count += 6
    return True

print(checkPrimeNumber(25))

False


# 95. Write a program to print all prime numbers between two given numbers a and b.


In [27]:
def getAllPrimes(num1, num2):
    primes = []
    for num in range(num1, num2+1):
        if num <= 1:
            continue
        if 1 < num < 4:
            primes.append(num)
        if num % 2 == 0 or num % 3 == 0:
            continue
        count = 5
        not_prime = False
        while count * count <= num:
            if (num % count == 0) or (num % (count + 2) == 0):
                not_prime = True
                break
            count += 6
        if not_prime:
            continue
        else:
            primes.append(num)
    return primes

print(getAllPrimes(1,1000))

[2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71, 73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151, 157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, 227, 229, 233, 239, 241, 251, 257, 263, 269, 271, 277, 281, 283, 293, 307, 311, 313, 317, 331, 337, 347, 349, 353, 359, 367, 373, 379, 383, 389, 397, 401, 409, 419, 421, 431, 433, 439, 443, 449, 457, 461, 463, 467, 479, 487, 491, 499, 503, 509, 521, 523, 541, 547, 557, 563, 569, 571, 577, 587, 593, 599, 601, 607, 613, 617, 619, 631, 641, 643, 647, 653, 659, 661, 673, 677, 683, 691, 701, 709, 719, 727, 733, 739, 743, 751, 757, 761, 769, 773, 787, 797, 809, 811, 821, 823, 827, 829, 839, 853, 857, 859, 863, 877, 881, 883, 887, 907, 911, 919, 929, 937, 941, 947, 953, 967, 971, 977, 983, 991, 997]


# 96. Write a program to find the sum of digits of a number repeatedly until a single digit remains.

1. digit_sum = 0
2. Take each digit from the given number, add it and store it in the variable 'digit_sum'
3. Check if the length of digit_sum is greater than 1; recall the same function using the same 'digit_sum' as input else, return that digit_sum,

In [8]:
def repeatedSumOfDigits(num):

    # Initialise the digit sum
    digit_sum = 0
    
    while num > 0:
        digit = num % 10
        digit_sum += digit
        num //= 10
        
    if len(str(digit_sum)) == 1:
        return digit_sum
    else:
        return repeatedSumOfDigits(digit_sum)

num = 12345 # => 1 + 2 + 3 + 4 + 5 = 15 => 1 + 5 = 6
print(repeatedSumOfDigits(num))
number2 = 9876 # => 9 + 8 + 7 + 6 = 30 => 3 + 0 = 3
print(repeatedSumOfDigits(number2))

6
3


# 97. Write a program to count the number of prime digits present in a number n.


In [11]:
def countPrimeDigits(number):
    prime_digits_lookup = [2,3,5,7]
    prime_digit_count = 0
    prime_digits_in_number = []
    while number > 0:
        digit = number % 10
        if digit in prime_digits_lookup:
            prime_digit_count += 1
            prime_digits_in_number.append(digit)
        number //= 10
    return prime_digit_count, prime_digits_in_number

n1 = 123456789
print(countPrimeDigits(n1))

(4, [7, 5, 3, 2])


# 98. Write a program to check whether a number is a palindrome and a prime at the same time.


In [16]:
def checkPalindrome(num):
    temp = num
    reversed_num = 0
    while temp > 0 :
        digit = temp % 10
        reversed_num = reversed_num * 10 + digit
        temp //= 10
    return reversed_num == num

def checkPrime(num):
    if num <= 1:
        return False
    if 1 < num < 4:
        return True
    if num % 2 == 0 or num % 3 == 0:
        return False
    count = 5
    while count * count <= num:
        if (num % count == 0) or ( num % (count + 2) == 0):
            return False
        count += 6
    return True

def checkBothPrimeAndPalindrome(num):
    return checkPrime(num) and checkPalindrome(num)

n1 = 111
print(checkPalindrome(n1))
print(checkPrime(n1))
print(checkBothPrimeAndPalindrome(n1))
n2 = 919
print(checkPalindrome(n2))
print(checkPrime(n2))
print(checkBothPrimeAndPalindrome(n2))
n3 = 29
print(checkPalindrome(n3))
print(checkPrime(n3))
print(checkBothPrimeAndPalindrome(n3))

True
False
False
True
True
True
False
True
False


# 99. Write a program to find the largest and smallest number that can be formed using the digits of n.


# 100. Write a program to convert a decimal number into its binary equivalent.


# 101. Write a program to convert a binary number into its decimal equivalent.


# 102. Write a program to display a menu that lets the user repeatedly choose any of the above tasks until they choose to exit.

# 103. Minimum steps to convert an Array into permutation of numbers from 1 to N (Using Addition/Subtraction)

In [1]:
def arr_permutation(arr):
    n = len(arr)

    # Sort the given array
    sorted_arr = sorted(arr)

    # minimum operations
    min_ops = 0

    for num in range(n):
        min_ops += abs(sorted_arr[num] - (num + 1))

    return min_ops

nums = [0, 2, 3, 4, 1, 6, 8, 9]
print(arr_permutation(nums))

7


# 104. Change the array into a permutation of numbers from 1 to n

In [5]:
def changeArrToPermutationByReplacing(arr):
    n = len(arr)
    freq_map = {}

    for num in arr:
        freq_map[num] = freq_map.get(num, 0) + 1

    min_ops = 0
    for num, count in freq_map.items():

        if num < 1 or num > n:
            min_ops += count
        else:
            min_ops += (count - 1)

    return min_ops

nums = [2,2,3,7] # need to replace 2 -> 1 and 7 -> 4 => output= 2
print(changeArrToPermutationByReplacing(nums))
nums1 = [0, 2, 3, 4, 1, 6, 8, 9] # need to replace 0 -> 5 and 9 -> 7 => output = 2
print(changeArrToPermutationByReplacing(nums1))

2
2
